MON-3257: Add Dockerfile.ocp and .ci-operator.yaml

slashpai opened the pull request.

slashpai committed:
Add Dockerfile.ocp

Signed-off-by: Jayapriya Pai <janantha@redhat.com>

slashpai committed:
Add .ci-operator.yaml

Signed-off-by: Jayapriya Pai <janantha@redhat.com>

slashpai commented:
/assign @simonpasquier

simonpasquier commented on Dockerfile.ocp:
```diff
@@ -0,0 +1,20 @@
+COPY . /go/src/sigs.k8s.io/metrics-server
+
+WORKDIR /go/src/sigs.k8s.io/metrics-server
+RUN make metrics-server
```
`make metrics-server` runs go build with the `-mod=readonly` flag which ignores the vendor directory (see https://go.dev/ref/mod#build-commands). Either we modify the Makefile command to have `-mod=vendor` or we expand the go build command here. The first option is probably the simpler as the Makefile also extrapolates the git tag, commit id, ... to pass them to the go build command.
Also I think that we need to remove the `.git` entry from the .dockerignore file otherwise the make command doesn't inject the `k8s.io/client-go/pkg/version.*` values.

jan--f replied:
Maybe a `Makefile.ocp` that does `include Makefile` and has a suitably modified `metrics-server-ocp` target?

Then we at least only have to worry about changes in the `metrics-server` target, but not about merge conflicts.

slashpai replied:
I was thinking if we can use this approach in syncbot https://github.com/rhobs/syncbot/blob/main/.github/workflows/merge-prometheus-adapter.yaml#L15-L16 so that we can keep downstream Makefile once we setup auto updates?

Either way is fine for me :)

jan--f replied:
Restoring our downstream version and overwriting possible future upstream changes to the Makefile worries me a bit. This lead to strange failure that will be hard to debug.

slashpai replied:
agree :)

Makefile.ocp

```
include Makefile

metrics-server-ocp: $(SRC_DEPS)
	GOARCH=$(ARCH) CGO_ENABLED=0 go build -mod=vendor -ldflags "$(LDFLAGS)" -o metrics-server sigs.k8s.io/metrics-server/cmd/metrics-server
```

we can update Dockefile.ocp to use this `make metrics-server-ocp -f Makefile.ocp`

simonpasquier commented on Dockerfile.ocp:
```diff
@@ -0,0 +1,20 @@
+
+WORKDIR /go/src/sigs.k8s.io/metrics-server
+RUN make metrics-server
+RUN mv /go/src/sigs.k8s.io/metrics-server/metrics-server /usr/bin/metrics-server
```
(nit) do we need this command? We can copy directly from /go/src/sigs.k8s.io/metrics-server/metrics-server in the final image?

slashpai committed:
Create new Makefile.ocp and update Dockerfile.ocp

Signed-off-by: Jayapriya Pai <janantha@redhat.com>

slashpai committed:
.dockerignore: Remove .git

This was needed so that docker doesn't
ignore k8s.io/client-go/pkg/version.* values

Signed-off-by: Jayapriya Pai <janantha@redhat.com>

danielmellado commented on Dockerfile.ocp:
```diff
@@ -0,0 +1,19 @@
+      summary="This is a component of OpenShift Container Platform that provides the Kubernetes resource metrics API based on Kubelet metrics" \
+      maintainer="OpenShift Monitoring Team <team-monitoring@redhat.com>"
+
+USER 1001
```
Just curious about why does the user or UID changes from the other provided Dockerfile, is there any specific reason for this? IIUC the user 65534 which is mapped in the other Dockerfile is the 'nobody' one, used for unprivileged access.

slashpai replied:
aah thats a good catch, I was referring to [prometheus-adapter ocp file](https://github.com/openshift/k8s-prometheus-adapter/blob/master/Dockerfile.ocp) for making this one so had used same UID. Looking at https://github.com/openshift/k8s-prometheus-adapter/blob/master/Dockerfile#L20 not sure if 1001 was used before updating in upstream. I will update to use `65534`

jan--f commented:
/lgtm
But for Dani's question and Simon to get another look on this
/hold

slashpai committed:
Update USER ID in Dockerfile.ocp

Signed-off-by: Jayapriya Pai <janantha@redhat.com>

slashpai commented:
@simonpasquier can you have another look to this one?

danielmellado commented:
/lgtm thanks jp!

jan--f commented:
/lgtm

simonpasquier reviewed (commented):
/lgtm
/hold cancel
